python/hail/expr/expressions/expression_typecheck.py

```diff
@@ -1,324 +1,455 @@
 
+    def can_coerce(self, t: HailType):
+        if self.elements is None:
+            return isinstance(t, ttuple)
+        else:
+            return (isinstance(t, ttuple)
+                    and len(t.types) == len(self.elements)
+                    and all(c.can_coerce(t_) for c, t_ in zip(self.elements, t.types)))
 
-def expr_array(t):
-    return ArrayChecker(lift(t))
+    def str_t(self):
+        if self.elements is None:
+            return 'tuple'
+        else:
+            return f'tuple({", ".join(c.str_t for c in self.elements)})'
 
+    def _coerce(self, x: Expression):
+        assert isinstance(x, hl.expr.TupleExpression)
+        return hl.tuple(c.coerce(element) for c, element in zip(self.elements, x))
 
-def expr_set(t):
-    return SetChecker(lift(t))
 
+class StructCoercer(ExprCoercer):
 
-def expr_dict(kt, vt):
-    return DictChecker(lift(kt), lift(vt))
+    def __init__(self, fields: Optional[Dict[str, ExprCoercer]] = None):
+        super(StructCoercer, self).__init__()
+        self.fields = fields
 
+    def _requires_conversion(self, t: HailType) -> bool:
+        assert isinstance(t, tstruct)
+        if self.fields is None:
+            return False
+        else:
+            return any(c._requires_conversion(t[name]) for name, c in self.fields.items())
 
-def expr_tuple(elements):
-    if elements is Ellipsis:
-        lifted = None
+    def can_coerce(self, t: HailType):
+        if self.fields is None:
+            return isinstance(t, tstruct)
+        else:
+            return (isinstance(t, tstruct)
+                    and len(t) == len(self.fields)
+                    and all(expected[0] == actual[0] and expected[1].can_coerce(actual[1])
+                            for expected, actual in zip(self.fields.items(), t.items())))
+
+    def str_t(self) -> str:
+        if self.fields is None:
+            return 'struct'
+        else:
+            field_strs = ', '.join(f'{escape_parsable(name)}: {c.str_t}' for name, c in self.fields.items())
+            return f'struct{{{field_strs}}})'
+
+    def _coerce(self, x: Expression):
+        assert isinstance(x, hl.expr.StructExpression)
+        assert list(x.keys()) == list(self.fields.keys())
+        return hl.struct(**{name: c.coerce(x[name]) for name, c in self.fields.items()})
+
+
+class OneOfExprCoercer(ExprCoercer):
+    def __init__(self, *options: ExprCoercer):
+        super(OneOfExprCoercer, self).__init__()
+        options_ = []
+        for o in options:
+            if isinstance(o, OneOfExprCoercer):
+                options_.extend(o.options)
+            else:
+                options_.append(o)
+        self.options = options_
+
+    @property
+    def str_t(self) -> str:
+        return ' or '.join(o.str_t for o in self.options)
+
+    def _requires_conversion(self, t: HailType) -> bool:
+        return all(o._requires_conversion(t) for o in filter(lambda c: c.can_coerce(t), self.options))
```

Comment:
I'm unsure about this. Can you explain the behavior you want in the various cases? If none of the options can be coerced, then `_requires_conversion` is True. Is that right?

Reply:
we know that can_coerce is True, so at least one of the `options` has can_coerce True. The type requires conversion if all of the options require conversion, because we want to take an identity route if it exists.

Reply:
Agreed.